supervisor/process.go

```diff
@@ -0,0 +1,411 @@
+		Mutex:        &sync.Mutex{},
+		stateLock:    &sync.Mutex{},
+		streamStdOut: stdOutChan,
+		streamStdErr: stdErrChan,
+		stdOut:       stdOut,
+		stdErr:       stdErr,
+	}
+
+	return process
+
+}
+
+// setState sets the state of a process
+func (p *Process) setState(state ProcessState) {
+	// cannot the set the state of a finished process
+	if p.state == Finished || p.state == state {
+		return
+	}
+
+	p.stateLock.Lock()
+	defer p.stateLock.Unlock()
+
+	// the next state can only be Starting if the state is Initial
+	if p.state == Initial {
+		p.state = Starting
+	} else {
+		p.state = state
+	}
+
+	p.status.State = p.state
+}
+
+// IsInitialState is true if process state is Initial
+func (p *Process) IsInitialState() bool {
+	p.stateLock.Lock()
+	defer p.stateLock.Unlock()
+
+	return p.state == Initial
+}
+
+// IsRunningState is true if process is Starting or Running
+func (p *Process) IsRunningState() bool {
+	p.stateLock.Lock()
+	defer p.stateLock.Unlock()
+
+	return p.state == Running || p.state == Starting
+}
+
+// IsFinalState is true if process is Finished or Interrupted
+func (p *Process) IsFinalState() bool {
+	p.stateLock.Lock()
+	defer p.stateLock.Unlock()
+
+	return p.state == Finished || p.state == Interrupted
+}
+
+// Start starts the process and returns a channel the caller can
+// use to retrieve the final status
+func (p *Process) Start() <-chan *Status {
+	p.Lock()
+	defer p.Unlock()
+
+	if p.statusChan != nil {
+		return p.statusChan
+	}
+
+	p.statusChan = make(chan *Status, 1)
+	p.setState(Starting)
+
+	p.cmd.Stdout = p.stdOut
+	p.cmd.Stderr = p.stdErr
+
+	p.cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
+
+	now := time.Now()
+	if err := p.cmd.Start(); err != nil {
+		// process failed to start.
+		// exit immediately
+		p.status.StartTime = now.UnixNano()
+		p.status.StopTime = time.Now().UnixNano()
+		p.status.Error = err
+		p.setState(Fatal)
+		p.statusChan <- &p.status
+		close(p.done)
+
+		return p.statusChan
+	}
+
+	p.status.StartTime = now.UnixNano()
+	p.status.PID = p.cmd.Process.Pid
+	p.setState(Running)
+	go p.wait()
+
+	return p.statusChan
+}
+
+// wait waits for a process to finish running and sets it's final status
+func (p *Process) wait() {
+	// send the final status and close done channel to signal to
+	// goroutines the process has finished running
+	defer func() {
+		p.statusChan <- p.GetStatus()
+		close(p.done)
+	}()
+
+	err := p.cmd.Wait()
+	now := time.Now()
+	exitCode := 0
+
+	p.Lock()
+	defer p.Unlock()
+
+	// get the exit code of the process
+	if err != nil {
+		if exitErr, ok := err.(*exec.ExitError); ok {
+			err = nil
+
+			if status, ok := exitErr.Sys().(syscall.WaitStatus); ok {
+				exitCode = status.ExitStatus()
+
+				if status.Signaled() {
+					err = errors.New(exitErr.Error())
+					p.setState(Interrupted)
+				}
+
+			}
+		}
+	}
+
+	// set final status of the process
+	p.status.Error = err
+	p.status.ExitCode = exitCode
+	p.status.StopTime = now.UnixNano()
+	p.setState(Finished)
```

Comment:
This will override Interrupted status if it was set above?

Reply:
Yes it overrides the Interrupted status

Reply:
What is the point of the `Interrupted` state then? Like `Starting`, it in immediately overwritten and never visible outside this function due to the lock.